File: java/src/televentas/Catalogo.java
package televentas;

import java.time.LocalDate;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Optional;

public class Catalogo {

    private final ArrayList<Producto> productos = new ArrayList<>();
    private LocalDate fechaActualizacion = LocalDate.now();

    public Catalogo() {
    }

    public Catalogo(List<Producto> inicial) {
        productos.addAll(inicial);
        fechaActualizacion = LocalDate.now();
    }

    public LocalDate getFechaActualizacion() {
        return fechaActualizacion;
    }

    public List<Producto> listarProductos() {
        return Collections.unmodifiableList(new ArrayList<>(productos));
    }

    public Optional<Producto> buscarProducto(String codigo) {
        for (Producto p : productos) {
            if (p.getCodigo().equals(codigo)) {
                return Optional.of(p);
            }
        }
        return Optional.empty();
    }

    public void agregarProducto(Producto producto) {
        if (buscarProducto(producto.getCodigo()).isPresent()) {
            throw new IllegalArgumentException("Ya existe un producto con codigo " + producto.getCodigo() + ".");
        }
        productos.add(producto);
        fechaActualizacion = LocalDate.now();
    }

    public void eliminarProducto(String codigo) {
        Optional<Producto> p = buscarProducto(codigo);
        if (!p.isPresent()) {
            throw new IllegalArgumentException("No existe un producto con codigo " + codigo + ".");
        }
        productos.remove(p.get());
        fechaActualizacion = LocalDate.now();
    }
}

File: javascript/src/televentas/catalogo.js
import { hoyUTC } from "./fechasUtil.js";

export class Catalogo {
  /**
   * @param {import('./producto.js').Producto[] | undefined} inicial
   */
  constructor(inicial) {
    /** @type {import('./producto.js').Producto[]} */
    this._productos = [];
    this.fechaActualizacion = hoyUTC();
    if (inicial != null && inicial.length > 0) {
      this._productos.push(...inicial);
      this.fechaActualizacion = hoyUTC();
    }
  }

  getFechaActualizacion() {
    return this.fechaActualizacion;
  }

  listarProductos() {
    return [...this._productos];
  }

  /** @returns {import('./producto.js').Producto | null} */
  buscarProducto(codigo) {
    for (const p of this._productos) {
      if (p.getCodigo() === codigo) {
        return p;
      }
    }
    return null;
  }

  agregarProducto(producto) {
    if (this.buscarProducto(producto.getCodigo()) != null) {
      throw new Error(`Ya existe un producto con codigo ${producto.getCodigo()}.`);
    }
    this._productos.push(producto);
    this.fechaActualizacion = hoyUTC();
  }

  eliminarProducto(codigo) {
    const p = this.buscarProducto(codigo);
    if (p == null) {
      throw new Error(`No existe un producto con codigo ${codigo}.`);
    }
    this._productos = this._productos.filter((x) => x.getCodigo() !== codigo);
    this.fechaActualizacion = hoyUTC();
  }
}
